src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffImageWriterBase.java

```diff
@@ -287,7 +330,20 @@ public void writeImage(final BufferedImage src, final OutputStream os, Map<Strin
         final int width = src.getWidth();
         final int height = src.getHeight();
 
-        int compression = TIFF_COMPRESSION_LZW; // LZW is default
+        final ColorModel cModel = src.getColorModel();
+        final boolean hasAlpha = cModel.hasAlpha() && checkForActualAlpha(src);
```

Comment:
Initially I thought maybe we didn't have to check for actual alpha in the image, since the colour model was already defining that it has alpha.

But this is for writing images, so it should be used just once... @gwlucastrig worst case scenario I guess would be a super large file, with no alpha. In that case We end up reading the whole image in memory (in chunks, but nevertheless...).

Do you think this could be a problem in the future? I'm fine with this change for now. If anybody reports any slowness we can just use a parameter to control whether this check is performed or not :man_shrugging: WDYT?

Reply:
The main reason I checked for actual alpha is that I'd seen a bunch of images of TYPE_INT_ARGB that had all the alpha values set to 0xff anyway.  I thought it would help reduce the size of the output file to leave it out if the image didn't actually use transparency.

Also, I've seen a bug in java's ImageIO class where if you try to store a BufferedImage of TYPE_INT_ARGB to a JPEG file, Java messes up the color.   That bug has been there like forever...  So I thought that it might avoid bugs for applications if by storing our TIFFs so that when they got read into BufferedImages by an application they would end up just simple TYPE_INT_RGB.  That way, an application that used Commons Imaging to read a TIFF and then ImageIO to write a JPEG might have better luck.

Reply:
Also, the ImageWriterBase always reads the whole image into memory.  That's one of my motivations for coming up with a new generation of image writer class in the JIRA item I posted recently.  The checkForActualAlpha has logic so that it reads a huge image in parts, so that limits the memory a little. Still, you are correct that it does mean two passes through the image data. But I think it is probably worth it to produce a smaller image on output.

Reply:
:+1: thanks for the detailed reply @gwlucastrig ! Is there some of what you said, that could go into a comment in this new code? For any new developers (or old ones with a horrible memory like me) of Imaging to be aware of why it's working this way.

Reply:
I've got some comments that I will incorporate into the code in my next PR.  Unfortunately, once again I've broken git and I can't get it to accept more changes to the current PR.  Clearly, there is something systematic I'm doing wrong and I am trying to figure out what that is.

In the meantime...  Here's the text I've written.  It would go into TiffImageWriterBase.java right before the call to checklForActualAlpha.

       // If the source image has a color model that supports alpha,
        // this module performs a call to checkForActualAlpha() to see whether
        // the image that was supplied to the API actually contains
        // non-opaque data in its alpha channel.  It is common for applications
        // to create a BufferedImage using TYPE_INT_ARGB, and fill the entire
        // image with opaque pixels. In such a case, the file size of the output
        // can be reduced by 25 percent by storing the image in an 3-byte RGB
        // format. This approach will also make a small reduction in the runtime
        // to read the resulting file when it is accessed by an application.

Reply:
@gwlucastrig I just had a look at your forked repo, and I believe I am able to push changes to your branch Gary. That's because by default pull requests allow contributors of a repository to push against the original branch of the pull request.

In that case, you should be able to push to your branch too. In a brand new directory, here's what I'd do:

```bash
git clone git@github.com:gwlucastrig/commons-imaging.git -o origin
git checkout issue-216
vim src/main/java/org/apache/commons/imaging/formats/tiff/write/TiffImageWriterBase.java 
git add .
git commit -m '[IMAGING-216] Updated documentation for TIFF writer base class'
git push -u origin issue-216
```

After that the new commit would appear here.

I used a new directory as that's what I'd have to do in my machine to push to your branch. But if you have a working directory with commons-imaging, you can try to use that to experiment with git and try to push to this PR.